File: Codechef/Long Contests/June Long 2019/forgotten_tree.cpp
/*
				C++
				encoding:UTF-8
				Modified: <13/Jun/2019 09:54:42 PM>

				✪ H4WK3yE乡
				Mohd. Farhan Tahir
				Indian Institute Of Information Technology (IIIT), Gwalior
*/

#include <bits/stdc++.h>

using namespace std;

// clang-format off

#define		 ll                     long long
#define		 mod                    1000000007
#define		 ve                     vector
#define		 pb                     push_back
#define		 umap                   unordered_map
#define		 pq                     priority_queue
//#define		 endl                   "\n";
#define		 ff                     first
#define		 ss                     second
#define		 pii                    pair<int, int>
#define		 len(v)                 int(v.size())
#define		 all(v)                 v.begin(), v.end()
#define		 reset(a, b)            memset(a, b, sizeof(a));
#define		 sum(v)                 accumulate(v.begin(), v.end(), (ll)0);
#define		 pr1(x)                 cout<<x<<"\n"
#define		 pr2(x, y)              cout<<x<<" "<<y<<"\n"
#define		 pr3(x, y, z)           cout<<x<<" "<<y<<" "<<z<<"\n"
#define		 prv(v, s, n)           fr(i, s, n) cout<<v[i]<<" "; cout<<"\n"
#define		 fr(i, s, n)            for (ll i = s ; i < n ; ++i)
#define		 dfr(i, s, n)           for (ll i = s ; i > n ; --i)

// clang-format on

signed main () {
	int tc;
	cin >> tc;
	while (tc--) {
		int n;
		cin >> n;
		int parent[n + 1];
		string s = "No";
		int curr = 1;
		int root;
		int lb = 1, rb = n;
		unordered_set< int > s;
		s.reserve (n + 1);
		s.max_load_factor (0.25);
		while (1) {
			cout << "Q " << curr << " " << lb << " " << rb << endl;
			cin >> s;
			if (s == "Yes") {
				parent[curr] = -1;
				root = curr;
				break;
			}
			curr++;
		}
	}
	return 0;
}

File: Codechef/Long Contests/June Long 2019/forgotten_tree.py
#    !/usr/bin/env python3
#    encoding: UTF-8
#    Modified: <14/Jun/2019 10:22:47 PM>


#    ✪ H4WK3yE乡
#    Mohd. Farhan Tahir
#    Indian Institute Of Information Technology (IIIT), Gwalior


def query(i, lb, ub):
    print('Q', i, lb, ub, flush=True)
    s = input()
    return True if s == 'Yes' else False


def main():
    from bisect import bisect_left as lower_bound
    from bisect import bisect_right as upper_bound
    from bisect import insort_left
    for tc in range(int(input())):
        t = 0
        n = int(input())
        arr = [0, n + 1]
        x = [-1] * (n + 2)
        parent = [0] * (n + 1)
        root = False
        while len(arr) != n + 2:
            for i in range(n, 0, -1):
                if i in arr:
                    continue
                if root == False:
                    lb, ub = 1, n
                else:
                    lb = lower_bound(arr, i)
                    lb -= 1
                    lb = arr[lb]
                    lb += 1
                    ub = upper_bound(arr, i)
                    ub = arr[ub]
                    ub -= 1

                if lb == i and root == True:
                    t += 1
                    x[i] = t
                    insort_left(arr, i)
                    lb = lower_bound(arr, i)
                    lb -= 1
                    ub = upper_bound(arr, i)
                    lb = arr[lb]
                    ub = arr[ub]
                    #print(lb, ub, x[lb], x[ub], arr[lb], arr[ub])
                    if x[lb] > x[ub]:
                        parent[i] = lb
                    else:
                        parent[i] = ub

                else:
                    ans = query(i, lb, ub)

                    if ans == True:
                        t += 1
                        x[i] = t
                        insort_left(arr, i)
                        if root == False:
                            parent[i] = -1
                            root = True
                        else:
                            lb = lower_bound(arr, i)
                            lb -= 1
                            ub = upper_bound(arr, i)
                            lb = arr[lb]
                            ub = arr[ub]
                            #print(lb, ub, x[lb], x[ub], arr[lb], arr[ub])
                            if x[lb] > x[ub]:
                                parent[i] = lb
                            else:
                                parent[i] = ub

                # print(arr)
        print("A", *parent[1:], flush=True)


if __name__ == "__main__":
    main()
